docs/reference/style-guides/node-style-guide.md

```diff
@@ -2,8 +2,311 @@
+
+Code should also follow the [Do's and Don'ts of TypeScript](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html).
+
+### Structure
+
+#### Organize by Feature
+
+(This is [present in the Google TypeScript Style Guide](https://google.github.io/styleguide/tsguide.html#organize-by-feature), but bears repeating)
+
+Code should be organized by feature, not by type.
+
+Don't:
+
+```
+app/
+├── controllers/
+│   └── auth.ts
+├── models/
+│   └── auth.ts
+├── validators/
+│   └── auth.ts
+```
+
+Do:
+
+```
+app/
+├── auth/
+│   ├── auth.controller.ts
+│   ├── auth.model.ts
+│   └── auth.validator.ts
+```
+
+If a feature has many DTOs, models, etc., a hybrid approach may be used, organized by feature first:
+
+```
+app/
+├── auth/
+│   ├── dto/
+│   ├── model/
+│   |   └── auth.model.ts
+│   ├── auth.controller.ts
+│   └── auth.validator.ts
+```
+
+Tests should be organized alongside the code being tested, with the `.spec.ts` suffix or `.in.spec.ts` for integration tests:
+
+```
+app/
+├── auth/
+│   ├── auth.controller.ts
+│   ├── auth.controller.spec.ts
+│   ├── auth.controller.in.spec.ts
+```
+
+#### NestJS File Naming Conventions
+
+##### File Naming for Class
+
+```typescript
+export class PascalCaseSuffix {} //= pascal-case.suffix.ts
+// Except for suffix, PascalCase to hyphen-case
+class FooBarNaming {} //= foo-bar.naming.ts
+class FooController {} //= foo.controller.ts
+// These are both DTO's, which the package namespace should reflect rather than
+// adding DTO to every class name and file name.
+class BarQueryInput {} //= dto/bar-query.input.ts
+class BarResultPayload {} //= dto/bar-result.payload.ts
+
+abstract class BaseAccount {} //= baseAccount.ts
+```
+
+#### Interface and Abstract Naming
+
+- Avoid prefixing with `I` or `Interface`, as this impacts readability and the type of object is separate from the name.
+- Don't include `Abstract` in the name of abstract classes, as this is the type. Prefer `Base` or `Default` prefix.
+
+```typescript
+interface User {}
+interface CustomUser extends User {}
+interface ThirdCustomUser extends CustomUser {}
+abstract class BaseAccount {}
+```
+
+#### Index Exporting
+
+1. It is recommended to place index.ts in each folder and export. Unless it's a special case, it is imported from a folder instead of directly from a file.
+
+Don't:
+
+```typescript
+import { FooController } from './feature/foo.controller';
+import { BarController } from './feature/bar.controller';
+```
+
+Do:
+
+```typescript
+import { FooController, BarController } from './feature';
+```
+
+2. Preferred method is to place only one file or folder name at the end of the path.
+
+Don't:
+
+```typescript
+import { UtilService } from '../common/providers/util.service';
+```
+
+Do:
+
+```typescript
+import { UtilService } from '../common';
+```
+
+### TypeScript
+
+#### Tests
```

Comment:
I think we should add a guideline concerning setup operations in tests. Essentially they need to inside beofore/Each/All blocks.

I recently cleaned up several tests that had setup that was not contained in a inside a before block. When setup is not inside a before/Each/All, the code is still executed when the testing framework evaluates the file. 

Consider the following :

Results in setupDB is called even though test is skipped.
```
const db = setupDB();
describe.skip('my db tests', () => {
   it('queries', () => {
      db.query('select * from foo');
   })
}
```

In some frameworks, this also results in setupDB is being being called even though test is skipped.
```
describe.skip('my db tests', () => {
   const db = setupDB();
   it('queries', () => {
      db.query('select * from foo');
   })
}
```

Good. setupDB is not called when test is skipped.
```
describe.skip('my db tests', () => {
  let db:DB;
  beforeAll(() => {
     db = setupDB();
  });
   it('queries', () => {
      db.query('select * from foo');
   })
}
```

Reply:
I see that there might be another 'best practices' doc, so maybe the above comment is outside the scope of the style guide.

Reply:
Yes, perhaps we want a larger document that goes into more detail purely around testing expectations, when one uses functional, integration, unit tests, etc. For example, when I worked on the Pyramid framework, we had this larger document:
https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/testing.html#testing-chapter

Something like that for our TypeScript back and front-end would be useful.